src/backends/perf_pt/decode.c

```diff
@@ -35,13 +35,317 @@
+ *
+ * Returns true on success or false otherwise. Upon failure, the start address
+ * is undefined.
+ */
+bool
+perf_pt_next_block(struct pt_block_decoder *decoder, int *decoder_status,
+                   uint64_t *addr) {
+    // If there are events pending, drain them first.
+    *decoder_status = drain_events(decoder, *decoder_status);
+    if (*decoder_status < 0) {
+        // Error.
+        return false;
+    } else if (*decoder_status & pts_eos) {
+        // End of stream.
+        *addr = 0;
+        return true;
+    } else if (*decoder_status > 0) {
+        // An unhandled case.
+        return false;
+    }
+
+    // Now fetch the block information.
+    struct pt_block block;
+    *decoder_status = pt_blk_next(decoder, &block, sizeof(block));
+    // Other +ve decoder status codes can arise here. We ignore them for now,
+    // and let them be detected by drain_events() above when we are next
+    // called.
+    if (*decoder_status < 0) {
+        if (*decoder_status == -pte_eos) {
+            // End of stream is flagged as an error in the case of pt_blk_next().
+            *addr = 0;
+            return true;
+        } else {
+            // A real error.
+            return false;
+        }
+    }
+
+    if (block.truncated == 1) {
+        // XXX A truncated block. This occurs when a block straddles a section
+        // boundary. In this case we may need some extra logic, but this should
+        // be a rare case.
+        return false;
+    }
+
+    if (block.ninsn == 0) {
+        // A block with zero instructions. Shouldn't happen.
+        return false;
+    }
+
+    *addr = block.ip;
+    return true;
+}
+
+/*
+ * Given a decoder and the last decoder status, drains events in the PT packet
+ * stream.
+ *
+ * The new decoder status is returned.
+ */
+static int
+drain_events(struct pt_block_decoder *decoder, int decoder_status) {
+    while(decoder_status & pts_event_pending) {
+        struct pt_event event;
+        decoder_status = pt_blk_event(decoder, &event, sizeof(event));
+        if (decoder_status < 0) {
+            return decoder_status;
+        }
+    }
+    return decoder_status;
+}
+
+/*
+ * Loads the libipt image `image` with the code of the current process.
+ *
+ * Returns true on success or false otherwise.
+ */
+static bool
+load_self_image(struct load_self_image_args *args)
+{
+    bool ret = dl_iterate_phdr(load_self_image_cb, args) == 0;
+    if (fsync(args->vdso_fd) == -1) {
+        ret = false;
+    }
+    return ret;
+}
+
+/*
+ * The callback for `load_self_image()`, called once for each program header.
+ *
+ * Returns 1 to stop iterating, and in our case to indicate an error. Returns 0
+ * on success and to continue iterating. See dl_iterate_phdr(3) for information
+ * on this interface.
+ */
+static int
+load_self_image_cb(struct dl_phdr_info *info, size_t size, void *data)
+{
+    ElfW(Phdr) phdr;
+    ElfW(Half) i;
+
+    (void) size;
```

Comment:
I'm not sure what this line is doing.

Reply:
It's a standard way to silence the compiler in the event of an unused argument. The argument is forced upon us by the signature of the `dl_iterate_phdr` callback.

Reply:
Aha! Maybe doc that with `// Unused variable: silence compiler warnings` or similar?

Reply:
I don't think we need to be quite that verbose, but I agree.